libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts

```diff
@@ -0,0 +1,352 @@
+        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
+        const id = typeof c !== "string" ? agColId(c) : c;
+        const result = this.headers.idToDescriptor[id];
+
+        invariant(result, `no column with id ${id}`);
+
+        return result;
+    }
+
+    /**
+     * Gets ColDef or ColGroupDef for column with the provided ID. This method will raise error if there is no such
+     * col: this is to draw out possible errors in the rest of table code that uses the cols and ColDefs.
+     *
+     * @param c - column id, Column, ColDef or ColGroupDef from ag-grid
+     */
+    public getColDef(c: string | Column | ColDef | ColGroupDef): ColDef | ColGroupDef {
+        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
+        const id = typeof c !== "string" ? agColId(c) : c;
+        const result = this.colDefs.idToColDef[id];
+
+        invariant(result, `no column with id ${id}`);
+
+        return result;
+    }
+
+    /**
+     * Gets descriptors of all attributes that are used to slice the table into rows. Note that it is perfectly
+     * OK that table has no slicing attributes.
+     *
+     * @returns empty if there are no slicing attributes
+     */
+    public getSlicingAttributes(): IAttributeDescriptor[] {
+        return this.headers.sliceCols.map((col) => col.attributeDescriptor);
+    }
+
+    /**
+     *
+     */
+    public sliceColCount(): number {
+        return this.headers.sliceCols.length;
+    }
+
+    /**
+     * Gets all descriptors of all measures that were used to compute data values for this table. Note that it is
+     * perfectly OK that table has no measures. If the table has slicing attributes, then it will be listing out
+     * all available attribute values.
+     *
+     * Also note that table MAY have more data leaf columns than there are number of measures. If the measures
+     * are further scoped for values of some attribute's elements, then there will be one data col for each combination
+     * of measure X attribute element.
+     *
+     * @returns empty if there are no measures
+     */
+    public getMeasures(): IMeasureDescriptor[] {
+        return this.dv.meta().measureDescriptors();
+    }
+
+    /**
+     * Returns count of leaf data cols. This represents the actual width of the data sheet holding the computed metric
+     * values.
+     */
+    public dataColLeavesCount(): number {
+        return this.dataColLeaves;
+    }
+
+    /**
+     * Tests whether the column with the provided id is the first (e.g. left-most) column in the table. Table always starts with
+     * slicing columns followed by data columns.
+     *
+     * @param c - column id, Column, ColDef or ColGroupDef from ag-grid
+     */
+    public isFirstCol(c: string | Column | ColDef | ColGroupDef): boolean {
+        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
+        const id = typeof c !== "string" ? agColId(c) : c;
```

Comment:
Nitpick: this serialization/sanitization is repeated several times in the class. Woudl a utility function be useful even outside of it? Or maybe just make `agColId` take strings as well?

Reply:
done